Introduce Admin API chapter (#1860)
PrestaShop/docs: admin-api/resource_server/api-resources.md

---
title: API Resources
menuTitle: API Resources
weight: 2
useMermaid: true
---------

# API Resources

> API Platform has many different ways to configure and define APIs (PHP, YAML, XML), we chose the PHP approach relying on PHP attributes to simplify and centralize the configuration in a single file which is why:

Reviewer: same paragraph than in the previous page in the `API Resource classes` paragraph?

Author: You're right I'll modify the previous page with a link to this one

> API Platform has many different ways to configure and define APIs (PHP, YAML, XML), we chose the PHP approach relying on PHP attributes to simplify and centralize the configuration in a single file which is why:

Reviewer 2: ```suggestion
API Platform has many different ways to configure and define APIs (PHP, YAML, XML). We chose the PHP approach relying on [PHP attributes](https://www.php.net/manual/en/language.attributes.overview.php) to simplify and centralize the configuration in a single file, which is why:
```

#[ApiResource(
    operations: [
        new PaginatedList(
            uriTemplate: '/api-clients',
            scopes: [
                'api_client_read',
            ],
            ApiResourceMapping: [
                '[id_api_client]' => '[apiClientId]',
                '[client_id]' => '[clientId]',
                '[client_name]' => '[clientName]',
                '[external_issuer]' => '[externalIssuer]',
            ],
            gridDataFactory: 'prestashop.core.grid.data_factory.api_client',
            filtersClass: ApiClientFilters::class,
            filtersMapping: [
                '[apiClientId]' => '[id_api_client]',
                '[clientId]' => '[client_id]',
                '[clientName]' => '[client_name]',
                '[externalIssuer]' => '[external_issuer]',
            ],
        ),
    ],
    normalizationContext: ['skip_null_values' => false],
)]
class ApiClientList
{
    #[ApiProperty(identifier: true)]
    public int $apiClientId;

    public string $clientId;

    public string $clientName;

    public string $description;

    public ?string $externalIssuer;

    public bool $enabled;

    public int $lifetime;
}
```

## Custom Mapping

> The API Platform DTO allows you to define the expected format of your API endpoint, including the way each field name format (snake case, camel case, ...) and any field you want to rename because it seems better.
> 
> But the CQRS or Grid implementation may not match the format or naming you expected, which is why you can define some mapping to explain how to match the fields between your DTO and the underlying implementation.
> 
> Each mapping is an associative array, the keys are the original naming and the value is the target mapping (the renamed key if you prefer), you can customize different mapping that are used in different moments of normalization in the workflow. 

Reviewer 2: ```suggestion
The API Platform DTO allows you to define the expected format of your API endpoint, including the format of each field name (snake case, camel case, etc.) and any field you want to rename because it seems better.

However, the CQRS or Grid implementation may not match the format or naming you expected, which is why you can define some mapping to explain how to match the fields between your DTO and the underlying implementation.

Each mapping is an associative array. The keys are the original naming, and the value is the target mapping (the renamed key if you prefer). You can customize different mappings that are used at different moments of normalization in the workflow. 
```
